### PR TITLE
Demo: Conditional fields, wrapped in object

### PR DESCRIPTION
**Pull request by Ben-Ho**
- using switch/toggle to hide/show conditional form fields with divider to separate to other fields
- all conditional fields are stored grouped in a single object
- initial state depends on existence of object
- disabling should keep values until saved
- conditional required fields should only be validated if enabled/visible

open
- should we add a demo for not-nested conditional fields?
- should we add a demo with form saving switch-state?
- label is above switch but should be to the left

https://github.com/vivid-planet/comet/assets/1324250/a8b432e6-2df9-472f-adb4-9fd2df8593cd

### CODE REVIEW

File: demo/admin/src/products/ManufacturerForm.tsx [hunk 1]
```diff
@@ -227,54 +232,72 @@ export function ManufacturerForm({ id }: FormProps): React.ReactElement {
...
+                                            ) : (
+                                                <FormattedMessage id="manufacturer.address.useAlternativeAddress.no" defaultMessage="No" />
+                                            )
+                                        }
+                                        control={<FinalFormSwitch {...props} />}
+                                    />
+                                )}
+                            </Field>
+                            {values?.useAlternativeAddress && (
+                                <>
```
Comment (dkarnutsch) [inline demo/admin/src/products/ManufacturerForm.tsx:256]: This can also be achieved by using Field twice, which is imho cleaner (see https://codesandbox.io/s/lm4p3m92q from the official examples)

File: demo/admin/src/products/ManufacturerForm.tsx [hunk 2]
```diff
@@ -227,54 +232,81 @@ export function ManufacturerForm({ id }: FormProps): React.ReactElement {
...
+                            <Field
+                                fullWidth
+                                name="useAlternativeAddress"
+                                type="checkbox"
+                                label={<FormattedMessage id="manufacturer.address.useAlternativeAddress" defaultMessage="Use alternative address" />}
+                            >
+                                {(props) => (
+                                    <FormControlLabel
+                                        label={
+                                            values?.useAlternativeAddress ? (
```
Comment (nsams) [inline demo/admin/src/products/ManufacturerForm.tsx:245]: Please don't introduce this "no"/"yes" based on the switch value in this PR. This is a new pattern (which I don't think we should use) that is not related to the conditional fields.

(And the interesting thing is that you implemented two different ways to access the useAlternativeAddress value)

Comment (Ben-Ho) [reply demo/admin/src/products/ManufacturerForm.tsx:245]: > Please don't introduce this "no"/"yes" based on the switch value in this PR. This is a new pattern (which I don't think we should use) that is not related to the conditional fields.

I didn't introduce this, I found it in the comet docs: https://storybook.comet-dxp.com/?path=/story/stories-form-finalform-fields--finalformswitch

> And the interesting thing is that you implemented two different ways to access the useAlternativeAddress value

If I had an idea how to combine the `<Field name="useAlternativeAddress" subscription={{` stuff with the `FormControlLabel`-label I would use the Field-subscription and remove the subscription on form-level.

Comment (nsams) [reply demo/admin/src/products/ManufacturerForm.tsx:245]: > I found it in the comet docs:

that looks like a dev story which is not part of the docs (yes, a confusing mix in our storybook)

Do you have a screendesign that has this label?

> If I had an idea how to combine

I'm not sure what your problem is, but can't you use `props.input.value`?

Comment (Ben-Ho) [reply demo/admin/src/products/ManufacturerForm.tsx:245]: > Do you have a screendesign that has this label?

image

> I'm not sure what your problem is, but can't you use props.input.value?

I'll try this.

Comment (Ben-Ho) [reply demo/admin/src/products/ManufacturerForm.tsx]: what about the '"no"/"yes" based on the switch value'? Should we keep it this way or do you have another idea how to achieve the same?

**Comment by johnnyomair** [comment]
There are some unwanted behaviors in the screencast (save conflict, form refreshing). Does this happen all the time, or only once?

File: demo/api/src/products/entities/manufacturer.entity.ts [hunk 3]
```diff
@@ -32,9 +32,9 @@ export class AlternativeAddress {
 @ObjectType()
 @InputType("AddressInput")
 export class Address extends AlternativeAddress {
+    @IsNullable()
     @Property({ type: "json", nullable: true })
     @Field(() => AlternativeAddress, { nullable: true })
-    @IsObject()
```
Comment (johnnyomair) [inline demo/api/src/products/entities/manufacturer.entity.ts:37]: Why do you remove this?

File: demo/admin/src/products/ManufacturerForm.tsx [hunk 4]
```diff
@@ -227,54 +232,81 @@ export function ManufacturerForm({ id }: FormProps): React.ReactElement {
...
+                                label={<FormattedMessage id="manufacturer.address.useAlternativeAddress" defaultMessage="Use alternative address" />}
+                            >
+                                {(props) => (
+                                    <FormControlLabel
+                                        label={
+                                            props.input.checked ? (
+                                                <FormattedMessage id="manufacturer.address.useAlternativeAddress.yes" defaultMessage="Yes" />
+                                            ) : (
+                                                <FormattedMessage id="manufacturer.address.useAlternativeAddress.no" defaultMessage="No" />
+                                            )
```
Comment (johnnyomair) [inline demo/admin/src/products/ManufacturerForm.tsx:249]: We have generic messages in `@comet/admin` for this: https://github.com/vivid-planet/comet/blob/next/packages/admin/admin/src/messages.ts#L31-L32

File: demo/admin/src/products/ManufacturerForm.tsx [hunk 5]
```diff
@@ -189,7 +194,7 @@ export function ManufacturerForm({ id }: FormProps): React.ReactElement {
             initialValuesEqual={isEqual} //required to compare block data correctly
             subscription={{}}
         >
-            {() => (
+            {({ values }) => (
```
Comment (johnnyomair) [inline demo/admin/src/products/ManufacturerForm.tsx:197]:
```suggestion
            {() => (
```

This isn't used

**Review by johnnyomair — COMMENTED** [review]
Looks good!

What's missing in your screencast is the demonstration of the required validation of the optional fields when the alternative address is used. Does this work as intended?

**Comment by Ben-Ho** [comment]
> There are some unwanted behaviors in the screencast (save conflict, form refreshing). Does this happen all the time, or only once?

save-conflict only occurs if saved to fast. (that's why I waited for the second save-click) And this is probably an issue with the save-conflict detection and not related to my changes.

form refreshing issues should be gone with the change from form-subscription to field-subscription, but I'll check this again.

File: demo/api/src/products/entities/manufacturer.entity.ts (hunk 3 above)
Comment (Ben-Ho) [reply demo/api/src/products/entities/manufacturer.entity.ts:37]: because this does not allow null-value.

**Comment by Ben-Ho** [comment]
> Looks good!
> 
> What's missing in your screencast is the demonstration of the required validation of the optional fields when the alternative address is used. Does this work as intended?

yes, it does.

**Comment by Ben-Ho** [comment]
> form refreshing issues should be gone with the change from form-subscription to field-subscription, but I'll check this again.

it's not gone, but it's the same with the current next-branch. So this should probably not be connected to any of the changes in this pull-request... It seems to be connected with the switch between add and edit. (it does not happen if saving an already existing entity)

File: demo/api/src/products/entities/manufacturer.entity.ts (hunk 3 above)
Comment (johnnyomair) [reply demo/api/src/products/entities/manufacturer.entity.ts:37]: What about using both validators together? Then it would be either null or object. Now you don't have any validation.

**Comment by johnnyomair** [comment]
> it's not gone, but it's the same with the current next-branch. So this should probably not be connected to any of the changes in this pull-request... It seems to be connected with the switch between add and edit. (it does not happen if saving an already existing entity)

I did some testing myself and it seem's that this is due to a rounding error in the updatedAt timestamps. I've started a discussion in our internal chat platform.

**Review by johnnyomair — DISMISSED** [review]

**Review by nsams — DISMISSED** [review]

**Review by johnnyomair — APPROVED** [review]
